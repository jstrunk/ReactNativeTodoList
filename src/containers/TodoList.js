/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 * @lint-ignore-every XPLATJSCOPYRIGHT1
 */

import React, {Component} from 'react';
import {Platform, StyleSheet, Text, View} from 'react-native';
import AsyncStorage from '@react-native-community/async-storage';
import FlashMessage, {showMessage, hideMessage} from 'react-native-flash-message';

import Title from '../components/Title';
import MainBody from '../components/MainBody';
import Footer from '../components/Footer';

type Props = {};

export default class TodoList extends Component<Props> {
  constructor(props) {
    super(props);
    this.state = {
      items: []
    };
  }

  componentDidMount() {
    this.loadData().done();
  }

  loadData = async () => {
    try {
      const items = await AsyncStorage.getItem('items');
      if (items !== null) {
        this.setState({items: JSON.parse(items)});
      } else {
        this.setState({items: []});
        this.storeData([]);
      }
    } catch (error) {
      showMessage({
        message: 'Error loading data',
        type: 'danger',
        autoHide: false,
      })
    }
  }

  storeData = async (items) => {
    try {
      await AsyncStorage.setItem('items', JSON.stringify(items));
    } catch (error) {
      // Error saving data
      showMessage({
        message: 'Error saving data',
        type: 'danger',
        autoHide: false,
      });
    }
  };

  onUpdate (items) {
    this.setState({items});
    this.storeData(items);
  }

  render() {
    return (
      <View style={styles.container}>
        <Title>Todo List</Title>
        <MainBody items={this.state.items} onUpdate={this.onUpdate.bind(this)} />
        <Footer items={this.state.items} onUpdate={this.onUpdate.bind(this)}>Remove completed items</Footer>
        <FlashMessage position="top" />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
  },
});
