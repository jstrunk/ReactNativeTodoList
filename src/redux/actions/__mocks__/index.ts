export const addItem = jest.fn();
export const toggleItem = jest.fn();
export const deleteItem = jest.fn(); 
export const deleteCompleted = jest.fn(); 
export const reorderList = jest.fn(); 
export const registerList = jest.fn(); 
export const setActiveList = jest.fn(); 
export const createList = jest.fn(); 
export const loadList = jest.fn(); 