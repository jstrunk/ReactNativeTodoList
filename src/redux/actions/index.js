import { ADD_ITEM, DELETE_ITEM, TOGGLE_ITEM, DELETE_COMPLETED } from '../constants';

export function addItem(text) {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index) {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index) {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted() {
  return { type: DELETE_COMPLETED };
}
