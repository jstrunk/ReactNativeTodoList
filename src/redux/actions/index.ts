import { ADD_ITEM,
  DELETE_ITEM,
  TOGGLE_ITEM,
  DELETE_COMPLETED,
  TodoActionType, } from '../types';

export function addItem(text: string): TodoActionType {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index: number): TodoActionType {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index: number): TodoActionType {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted(): TodoActionType {
  return { type: DELETE_COMPLETED };
}
