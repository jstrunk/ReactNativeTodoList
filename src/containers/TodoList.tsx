/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React from 'react';
import { connect, ConnectedProps } from 'react-redux';
import List from '../components/List';
import { reorderList } from '../redux/actions';
import { ITodoState } from '../redux/types';
import { ThunkDispatch } from 'redux-thunk';
import { Text, View, ActivityIndicator } from 'react-native';
import styles from '../styles';

const mapStateToProps = (state: ITodoState) => ({
  todoList: state.todoList,
})

const mapDispatchToProps = (dispatch: ThunkDispatch<any, any, any>) => {
  return {
    reorderList: (ids: Array<string>) => {
      dispatch(reorderList(ids));
    }
  }
};

const connector = connect(
  mapStateToProps,
  mapDispatchToProps,
);

type PropsFromRedux = ConnectedProps<typeof connector>;

const TodoList = (props: PropsFromRedux) => {
  switch (props.todoList.status) {
    case 'fetching':
      return (
        <View style={styles.centeredView}>
          <ActivityIndicator size='large' />
        </View>
      );
    case 'error':
      return (
        <View style={styles.centeredView}>
          <Text>Failed to load list.</Text>
        </View>
      );
    case 'loaded':
    default:
      return (<List items={props.todoList.allItems} reorderList={props.reorderList} />);
  }
}

export default connector(TodoList);
