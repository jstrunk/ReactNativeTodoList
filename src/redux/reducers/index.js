import { combineReducers } from 'redux';
import listReducer from '../reducers/listReducer';

export default combineReducers(
  { todoList: listReducer }
);