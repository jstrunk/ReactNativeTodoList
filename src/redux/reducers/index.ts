import { combineReducers } from 'redux';
import listReducer from '../reducers/listReducer';
import itemReducer from '../reducers/itemReducer';
import { ITodoState } from '../types';

export default combineReducers<ITodoState>(
  {
    todoList: listReducer,
    todoItems: itemReducer,
  }
);
