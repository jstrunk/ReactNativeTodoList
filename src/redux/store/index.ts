import AsyncStorage from '@react-native-community/async-storage';
import { createStore, Store } from 'redux';
import { persistStore, persistReducer, Persistor } from 'redux-persist';
import rootReducer from '../reducers';

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

export interface PersistStore {
  store: Store,
  persistor: Persistor,
}
const configureStore = (): PersistStore => {
  const store = createStore(persistedReducer);
  const persistor = persistStore(store);
  return { store, persistor };
}

export default configureStore;
