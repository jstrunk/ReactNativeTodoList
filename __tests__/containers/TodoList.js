import 'react-native';
import React from 'react';
import TodoList from '../../src/containers/TodoList';
import * as actions from '../../src/redux/actions';
import { Provider } from 'react-redux';
import { render, fireEvent, screen } from '@testing-library/react-native'
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);

describe('TodoList container', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      todoList: [
        {
          key: 'Run the tests',
          done: true,
        },
        {
          key: 'foo',
          done: false,
        },
      ],
    });

    store.dispatch = jest.fn();
  });

  it('renders a TodoList', () => {
    const { asJSON } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );
    expect(asJSON()).toMatchSnapshot();
  });

  it('should dispatch a toggle item action', () => {
    const { getAllByText } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    const checkedElements = getAllByText('');
    expect(checkedElements).toHaveLength(1);

    checkedElements.forEach((e) => fireEvent.press(e));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.toggleItem(0)
    );
  });

  it('should dispatch a delete item action', () => {
    const { getAllByText } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    const deleteElements = getAllByText('X');
    expect(deleteElements).toHaveLength(2);

    deleteElements.forEach((e) => fireEvent.press(e));

    expect(store.dispatch).toHaveBeenCalledTimes(2);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.deleteItem(0)
    );
  });
});