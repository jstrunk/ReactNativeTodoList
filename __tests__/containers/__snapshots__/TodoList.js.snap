// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList container renders a TodoList 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "done": true,
          "key": "Run the tests",
        },
        Object {
          "done": false,
          "key": "foo",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    maxToRenderPerBatch={10}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={Object {}}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={null}
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "black",
                    "fontSize": 17,
                    "textAlign": "left",
                  }
                }
                testID="listItemTitle"
                theme={
                  Object {
                    "colors": Object {
                      "disabled": "hsl(208, 8%, 90%)",
                      "divider": "#bcbbc1",
                      "error": "#ff190c",
                      "grey0": "#393e42",
                      "grey1": "#43484d",
                      "grey2": "#5e6977",
                      "grey3": "#86939e",
                      "grey4": "#bdc6cf",
                      "grey5": "#e1e8ee",
                      "greyOutline": "#bbb",
                      "platform": Object {
                        "android": Object {
                          "error": "#f44336",
                          "primary": "#2196f3",
                          "secondary": "#9C27B0",
                          "success": "#4caf50",
                          "warning": "#ffeb3b",
                        },
                        "ios": Object {
                          "error": "#ff3b30",
                          "primary": "#007aff",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                      },
                      "primary": "#2089dc",
                      "searchBg": "#303337",
                      "secondary": "#8F0CE8",
                      "success": "#52c41a",
                      "warning": "#faad14",
                    },
                  }
                }
              >
                Run the tests
              </Text>
            </View>
            <View
              width={16}
            />
            <TouchableOpacity
              checkedIcon="check-square-o"
              size={24}
              style={
                Object {
                  "margin": 0,
                  "marginLeft": 0,
                  "marginRight": 0,
                  "padding": 0,
                }
              }
              testID="checkbox"
              uncheckedColor="#bfbfbf"
              uncheckedIcon="square-o"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#2089dc",
                        "fontSize": 24,
                      },
                      Object {
                        "minWidth": 24,
                      },
                      Object {
                        "fontFamily": "FontAwesome",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
              </View>
            </TouchableOpacity>
            <View
              width={16}
            />
            <TouchableOpacity>
              <Text
                style={
                  Object {
                    "color": "red",
                    "textAlign": "center",
                  }
                }
              >
                X
              </Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>
      <View
        style={null}
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "black",
                    "fontSize": 17,
                    "textAlign": "left",
                  }
                }
                testID="listItemTitle"
                theme={
                  Object {
                    "colors": Object {
                      "disabled": "hsl(208, 8%, 90%)",
                      "divider": "#bcbbc1",
                      "error": "#ff190c",
                      "grey0": "#393e42",
                      "grey1": "#43484d",
                      "grey2": "#5e6977",
                      "grey3": "#86939e",
                      "grey4": "#bdc6cf",
                      "grey5": "#e1e8ee",
                      "greyOutline": "#bbb",
                      "platform": Object {
                        "android": Object {
                          "error": "#f44336",
                          "primary": "#2196f3",
                          "secondary": "#9C27B0",
                          "success": "#4caf50",
                          "warning": "#ffeb3b",
                        },
                        "ios": Object {
                          "error": "#ff3b30",
                          "primary": "#007aff",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                      },
                      "primary": "#2089dc",
                      "searchBg": "#303337",
                      "secondary": "#8F0CE8",
                      "success": "#52c41a",
                      "warning": "#faad14",
                    },
                  }
                }
              >
                foo
              </Text>
            </View>
            <View
              width={16}
            />
            <TouchableOpacity
              checkedIcon="check-square-o"
              size={24}
              style={
                Object {
                  "margin": 0,
                  "marginLeft": 0,
                  "marginRight": 0,
                  "padding": 0,
                }
              }
              testID="checkbox"
              uncheckedColor="#bfbfbf"
              uncheckedIcon="square-o"
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#bfbfbf",
                        "fontSize": 24,
                      },
                      Object {
                        "minWidth": 24,
                      },
                      Object {
                        "fontFamily": "FontAwesome",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  
                </Text>
              </View>
            </TouchableOpacity>
            <View
              width={16}
            />
            <TouchableOpacity>
              <Text
                style={
                  Object {
                    "color": "red",
                    "textAlign": "center",
                  }
                }
              >
                X
              </Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
